### PR TITLE
[CI][Doc] Add design proposal for make master stable task

### PR DESCRIPTION
**Pull request by 119Vik**
Signed-off-by: Vitalii Kostenko <vitalii@freedomfi.com>

## Summary
Add design proposal for #3564 .

### CODE REVIEW

File: docs/readmes/proposals/p006_mandatory_integration_tests_for_each_PR.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+- Reviewer did not wait for tests
+- Two conflicting PRs were merged that had logic that breaks something
+- External factors not covered by PR process (such as a library package update)
+This regression pipeline runs twice a day(EU time / US night time) on a schedule. If the build fails, a GitHub issue is created tagging Magma CI team for urgent attention. 
+
+### Magma_integration_tests pipeline is the largest bottleneck in our CI process
+At the moment magma_integration_tests pipeline is structured in such a way that it's using a Jenkins slave as a jump host to a real test environment. We only have one real test env for the magma_integration_tests for now. We need to do some simple refactoring to ensure that magma_integration_tests utilizes a Jenkins slave as a real test env (like CWAG and LTE integration tests do). This means replacing all the 'ssh <user>@<host> - '<command>'' with corresponding 'sh <command> ' statements in the pipeline definition. This simple refactoring should allow Jenkins to handle up to 16 builds per hour (instead of just 1 at the present moment).
+
+## Executio plan
```
Comment (mattymo) [inline docs/readmes/proposals/p006_mandatory_integration_tests_for_each_PR.md]: Execution

**Review by karthiksubraveti — COMMENTED** [review]
Thanks @119Vik for suggesting this. It sounds like a good idea. Stage 2 will be more convenient. Can you also add a section on what we plan to do when we have flaky integration tests or any bad integration test code merges which block all the PRs from merging and how we can handle thundering herds problem in such event where all PRs will merge and break the master again.

File: docs/readmes/proposals/p006_mandatory_integration_tests_for_each_PR.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ - Ensure that each PR marked with label "ready-to-merge" was tested by all integration tests.
+ Stage 2:
+ - Ensure that each PR marked with label "ready-to-merge", which passed integration tests is automatically merged by CI.
+ Stage 3:
+ - Ensure that no side effects appear when PRs are automatically merged by CI (so speculative CI implementation in fact).
+
+### Some PRs are still merged into master without integration tests
+Make Jenkins as a mandatory gatekeeper for merging each PR. Proposed workflow:
+ - Reviewer marks appropriate labels to run necessary relevant tests (such as CWAG / NMS / etc.) before indicating "ready-to-merge" (tag or some keyword in a comment)
+ - Jenkins triggers all tests (or if we decide, just a subset)
```
Comment (amarpad) [inline docs/readmes/proposals/p006_mandatory_integration_tests_for_each_PR.md]: We should start with all integration tests and then try and understand why we will reduce the surface area if needed.

**Review by amarpad — COMMENTED** [review]
Thanks for this, looks good to me.

For the flakey tests, we should just file hi-pri issues and disable them aggressively.

**Comment by 119Vik** [comment]
@karthiksubraveti - thanks.
>  flaky integration tests

 to distinguish buggy code from flaky tests we need to do case review. As well as flaky test is kind of a bug as well. PR with Fix should be prepared ASAP in any case.

> any bad integration test code merges which block all the PRs

idea of this proposal and discussion is to eliminate possibility buggy code being merged into the master.

> how we can handle thundering herds problem in such event where all PRs will merge and break the master again

this is one of the problems speculative CI should fix - cause it implements sequential merge process (for each PR or groups of PRs) - before merge PR(s) should be rebased on the lates master and verified by integration tests. As a result each PR or butch of PRs is verified against latest code right before merge.

File: docs/readmes/proposals/p006_mandatory_integration_tests_for_each_PR.md (hunk 2 above)
Comment (119Vik) [reply docs/readmes/proposals/p006_mandatory_integration_tests_for_each_PR.md]: make sense - removed this comment

**Comment by ttx** [comment]
Looks good to me, sounds like this plan should speed up things and guarantee nothing bypasses tests. I'd advise completing the first two steps (refactoring and nightly builds) before switching to gating (mandatory testing): tests need to run reasonably fast before gating can be accepted by developers.

Happy to present Zuul speculative CI features when you are ready to consider it, but I would recommend implementing all the other steps first (as I suspect they will solve 99% of the issues). Note that if you implement speculative CI gating, you will no longer need to run nightly builds to catch merge-time regressions.

**Comment by 119Vik** [comment]
@ttx @amarpad @tmdzk - I've just updated the proposal with few more items which should solve performance issue we found recently.

**Comment by tmdzk** [comment]
@mattymo @119Vik @amarpad  should we merge/Close this? 
I suggest as it's an open discussion that is not fixed yet, we move it to github discussion. What do you guys think?

**Pull request merged**